### PR TITLE
Cosmos: Add support for executing pre- and post-triggers

### LINKED ISSUE #19944: CosmosDb + Triggers enabled
It is my understanding that CosmosDb has no way to turn triggers on by default. They have to be turned on per call.  EF Core doesn't turn these on, so triggers can't be used with EF Core. While I understand the separate request for the change feed, I'm referring to built in triggers.

This prevents basic house keeping operations on nosql databases because of no joins. Typically you have data that is a subset of a referenced entity.  This happens everywhere, but without triggers you can't update all of those references with the new subset of data of any of those fields change in the database.  I.e. if I have a contact, and it's referenced somewhere else, I might include the Name and Email of that contact in some other container that has has a contact associated with it so that you can query and pull all of the information needed about that contact for display without having to separately query for the contacts and join it in memory.  

If you have triggers, you can setup your contacts table to, by default, update all of the references in all other containers/objects when those fields change and update the reference information as necessary.

Without them, you have to do this logic in your C# logic every time you update the contact in every single case which is non-optimal and causes a major amount of maintenance code and risk in your C# code that should be managed by the datastore itself.

I would like EF Fluent instructions to allow turning on triggers and defining the triggers that are executed by default.  These might be Azure Function triggers, or they could be native cosmosdb triggers, doesn't really matter, we just need to be able to turn them on.
Further there should be an extension method that allows disabling them temporarily for batch operations etc.

Without this functionality you're not properly separating data store functions from data access functions.

### PR DESCRIPTION
**Pull request by AndriySvyryd**
Fixes #19944

### CODE REVIEW

File: src/EFCore.Cosmos/Extensions/CosmosTriggerBuilderExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+/// </remarks>
+public static class CosmosTriggerBuilderExtensions
+{
+    /// <summary>
+    ///     Configures the trigger type (Pre or Post).
+    /// </summary>
+    /// <param name="triggerBuilder">The builder for the trigger being configured.</param>
+    /// <param name="triggerType">The trigger type to configure.</param>
+    /// <returns>The same builder instance so that multiple configuration calls can be chained.</returns>
+    public static TriggerBuilder HasTriggerType(this TriggerBuilder triggerBuilder, TriggerType triggerType)
```
Comment (AndriySvyryd) [inline src/EFCore.Cosmos/Extensions/CosmosTriggerBuilderExtensions.cs]: This introduces public methods, so API review needs to be done as part of the PR review.
Note that `TriggerType` and `TriggerOperation` are from Cosmos SDK

**Comment by SamMonoRT** [comment]
cc fyi @Pilchie

**Comment by Pilchie** [comment]
Also tagging @kirankumarkolli .

File: test/EFCore.Cosmos.FunctionalTests/CosmosTriggersTest.cs [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+            modelBuilder.Entity<Product>(entity =>
+            {
+                entity.Property(e => e.Id);
+                entity.HasPartitionKey(e => e.PartitionKey);
+                entity.HasTrigger("PreInsertTrigger")
+                    .HasTriggerType(TriggerType.Pre).HasTriggerOperation(TriggerOperation.Create);
+                entity.HasTrigger("PostDeleteTrigger")
+                    .HasTriggerType(TriggerType.Post).HasTriggerOperation(TriggerOperation.Delete);
+                entity.HasTrigger("UpdateTrigger")
+                    .HasTriggerType(TriggerType.Pre).HasTriggerOperation(TriggerOperation.Replace);
```
Comment (AndriySvyryd) [inline test/EFCore.Cosmos.FunctionalTests/CosmosTriggersTest.cs]: Here's an example of the new API usage

Comment (roji) [reply test/EFCore.Cosmos.FunctionalTests/CosmosTriggersTest.cs]: Aren't the trigger type and operation mandatory? If so, I'd expect HasTrigger to have them as required parameters, no?

I get that there will be edge cases where someone wants to change the type/operation for a trigger that has already been defined, but in the 99% case users just want to define a new trigger (it might be acceptable for the edge case to just use the metadata APIs directly).

File: src/EFCore.Cosmos/Storage/Internal/CosmosClientWrapper.cs [hunk 3]
```diff
@@ -419,6 +420,20 @@ private static async Task<bool> CreateItemOnceAsync(
         var container = wrapper.Client.GetDatabase(wrapper._databaseId).GetContainer(parameters.ContainerId);
         var itemRequestOptions = CreateItemRequestOptions(entry, wrapper._enableContentResponseOnWrite);
         var partitionKeyValue = ExtractPartitionKeyValue(entry);
+        var preTriggers = GetTriggers(entry, TriggerType.Pre, TriggerOperation.Create);
```
Comment (roji) [inline src/EFCore.Cosmos/Storage/Internal/CosmosClientWrapper.cs:423]: Am noting that when performing an update using the Cosmos SDK, users need to manually specify which triggers they want executed; this is very different from relational database triggers, where the trigger is defined on the table and then automatically (and always) runs when the table is updated, without any client request.

This EF implementation seems to provide relational-like behavior, i.e. the user defines the trigger on the model and then the trigger gets implicitly used in SaveChanges (a more Cosmos-faithful implementation would allow the user to specify triggers to be executed on each SaveChanges). FWIW the relational behavior generally makes more sense to me as a developer experience, but maybe we should confirm this proposed behavior with the Cosmos people? i.e. maybe there's some reason we're missing for wanting users to manually specify triggers?

File: test/EFCore.Cosmos.FunctionalTests/CosmosTriggersTest.cs [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+
+            await context.SaveChangesAsync();
+
+            var logs = await context.Set<TriggerExecutionLog>().Where(l => l.Operation == "DELETE").ToListAsync();
+            
+            Assert.Contains(logs, l => l.TriggerName == "PostDeleteTrigger" && l.Operation == "DELETE");
+        }
+    }
+
+    private async Task CreateTriggersInCosmosAsync(TriggersContext context)
```
Comment (roji) [inline test/EFCore.Cosmos.FunctionalTests/CosmosTriggersTest.cs:64]: Don't we want EF to create the triggers with the Cosmos SDK? i.e. allow the user to configure the Javascript in the model and then do the CreateTriggerAsync call?

**Comment by Pilchie** [comment]
And tagging @TheovanKraay too.

File: test/EFCore.Cosmos.FunctionalTests/CosmosTriggersTest.cs (hunk 2 above)
Comment (AndriySvyryd) [reply test/EFCore.Cosmos.FunctionalTests/CosmosTriggersTest.cs]: Makes sense. I'll combine them into one method here, but leave separate on the Metadata API

File: src/EFCore.Cosmos/Storage/Internal/CosmosClientWrapper.cs (hunk 3 above)
Comment (AndriySvyryd) [reply src/EFCore.Cosmos/Storage/Internal/CosmosClientWrapper.cs:423]: I'd be interested to hear thoughts on this from the Cosmos team. But from what I've seen online this is the behavior that most users are looking for.
Also, from a technical perspective, we don't have a way of flowing provider-specific parameters through SaveChanges. If we did, we could add a way of opting out for specific triggers.

File: test/EFCore.Cosmos.FunctionalTests/CosmosTriggersTest.cs (hunk 4 above)
Comment (AndriySvyryd) [reply test/EFCore.Cosmos.FunctionalTests/CosmosTriggersTest.cs:64]: That's out of scope for this PR. Also, without schema versioning, this would not be useful/recommended in production

Comment (AndriySvyryd) [reply test/EFCore.Cosmos.FunctionalTests/CosmosTriggersTest.cs:64]: Filed #36524

File: src/EFCore.Cosmos/Storage/Internal/CosmosClientWrapper.cs (hunk 3 above)
Comment (AndriySvyryd) [reply src/EFCore.Cosmos/Storage/Internal/CosmosClientWrapper.cs:423]: Filed #36525

**Review by roji — APPROVED** [review]

**Comment by roji** [comment]
Oops this had auto-merge :(
